Gluon-milestone_1

kevingzhang opened the pull request:
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/8Wx7nxtq8fKrsuEz8) has been filled out for this milestone. 
- [x] This pull request is being made by the same account as the accepted application. 
- [x] In case of acceptance, the payment will be transferred to the initial BTC/ETH payment address. 
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/General-Grants-Program/blob/master/grants/milestone-deliverables-guidelines.md).

Noc2 commented:
Thanks for the delivery. We will look into it as soon as possible.

alxs commented:
Hi @kevingzhang. Would you mind adding the [tasks](https://github.com/w3f/Open-Grants-Program/blob/master/applications/Gluon_decentralized_hardware_crypto_wallet_services.md#milestone-1-example--web-app-and-mobile-app-pairing) for milestone 1 from the application into your delivery, and could you read through the items on your first comment and tick the checkmarks?

felixniemeyer commented:
Hey there @kevingzhang, 

I am trying to walk through what you did in the video on my computer but I cannot get the mobile app to work easily. 

Running the web app works fine and I get to the point where I could scan the QR code for pairing a mobile device. 

For continuing from there, I would need to run the mobile app somehow. 
I dug out an old iPhone 5s on which the iOS app provided by download from zqapps.com seems stuck at the end of the installation process.  The icon on my home screen is greyed out and the app cannot be started. 

I then tried building the android app. `npm run android` failed with `Could not initialize class org.codehaus.groovy.runtime.InvokerHelper`  and I resolved this error by downgrading jdk to 11 or switching to gradle 6.3 from gradle 6.2. Maybe you can put build environment specs in the readme? 
 
I could resolve the next error (`Could not determine the dependencies of task ':app:mergeDebugAssets'.`) by adding a `missingDimensionStrategy` to the `build.gradle` config. 

But then a new error arises (`Execution failed for task ':app:mergeDexDebug'.`) so I ask myself whether building for android is even meant to work at the moment.

I haven't tried emulating an iOS device because I am running linux and tools like `pod` from the instructions in the readme are not available. 

Can you help me get the mobile app running, so that I can try out the full walkthrough from the video?

kevingzhang commented:
> Hi @kevingzhang. Would you mind adding the [tasks](https://github.com/w3f/Open-Grants-Program/blob/master/applications/Gluon_decentralized_hardware_crypto_wallet_services.md#milestone-1-example--web-app-and-mobile-app-pairing) for milestone 1 from the application into your delivery, and could you read through the items on your first comment and tick the checkmarks?

Done!

kevingzhang commented:
Hi @felixniemeyer,

Thank you for reviewing our deliverables.

My apologies that we missed a few important items in our readme. We do not have an android version in milestone 1. We only tested on iPhone and ios simulator on a Mac. Android version is planed in milestone 3 before the General Release since the logic is the same and the dev framework is the same. 
We did not use testflight at this moment. The easiest way is to run it in a Mac under an ios simulator. If you do not have a Mac, there is another way that we used to test. You can follow the new updated readme.md to get your iPhone's UUID. Send the UUID to our mobile developer (his email is listed in the readme). He will repack the mobile app with your UUID inside, republish to the URL so that you can download and test it on your own iPhone. 

We have not tested an old iPhone yet. Technically our app should run on iPhone 8 or later. We tested on iPhone X and newer versions. If you do not have a newer iPhone, we would suggest you use an ios simulator.

Given ios simulator cannot scan QR code, we have a testing feature designed for this kind of testing process. You can copy and paste the content of the QR code to the ios simulator. This feature is for testing only. It will be removed from our final product. 

Please let me know which approach you prefer. If you choose to send us your UUID, please wait for a newer build to be uploaded to the URL so that you can continue testing. 

Thank you very much.

kevingzhang commented:
@felixniemeyer We just updated the milestone1 branch. We added the Android version mobile app. If you do not want to try the ios version, you can try android. It could be easier without dealing with TestFlight. 

Please let me know if you still have any issues with the testing.

Thanks

felixniemeyer commented:
Hi @kevingzhang 
awesome!
I have already run it on my phone and pairing worked. 
I'm planning to finish the external review tomorrow.

felixniemeyer commented:
Hi @kevingzhang,
I have found all of the milestone 1 features to work. There are a couple of small issues. You can find the full review here:  https://github.com/w3f/Grant-Milestone-Delivery/blob/bdaf1dcb2c6075dfc628b6d8f6c2ab1b7c2bdcf2/evaluations/gluon-1-felixniemeyer.md
I hope this is valuable feedback for you!
Kind regards

mmagician commented:
@kevingzhang It would be great if you could address all issues mentioned in @felixniemeyer 's review, including updating the READMEs to allow anyone to launch your frontend as well as build the mobile apps locally. Thanks!

kevingzhang commented:
All repositories have been updated according to @felixniemeyer 's reviews.

Some notes:

0. Readme has been added. The UI is not the final version. It is for API testing only. The UI will be redesigned in our following 4 milestones. You can see a lot of "test" buttons and tabs in our current webapp. All of them will be gone later. 
1. Pairing is the only feature in milestone 1. That's why we did not mention all other features since they are not available in milestone 1.
2. Added building android mobile app instruction in readme
3. Added the API source code location in the task list so that reviewers can easily locate the API function in the source code
4. Regarding the "search UI" question (task1/1). There is no UI for search. It is an API used by the web/mobile app so search for the current user's profile and show it on the UI. 
5. Question on task 2.2. Answer: Yes, those are all that the user profile contains in milestone1. In our later milestone, there would be more properties added to the user profile
6. Question on task 3.1. Answer: The API source code location has been added to the Deliverable application. 
7. Typo corrected in Deliverable application
8. Use of "lodash". Performance fine-tune is planned in milestone 4 and 5. In milestone1 we are OK to use lodash given the developers are familiar with this "old" style. All of these kinds of refactorings are planed for future milestones
9. Rust nightly build. Given Rust compiler upgrade very fast, we are using a fixed version in development. I updated the readme file to explain so that other reviewers do not need to bother the rust nightly version. 

Thank you very much for all the detailed reviews.

mmagician commented:
Thanks for the updates, though I think you still missed the following points from @felixniemeyer 's review:
> 1. There is a lot of code duplication in gluon-app/mobile/layer1/[gluon, rpc, ...].js and gluon-app/webapp/src/tea/[gluon, rpc, ...].js.

Could you possibly refactor the shared code into a `common` module that you re-use both in webapp and the mobile app?

---
> 2. It seems to me, that you can remove the template pallet in gluon-layer1.

---
Furthermore, some notes from my side:
1. You're assuming that the gluon pallet is located in `../../gluon-pallet/gluon`, which depends on the users' setup (I haven't seen instructions to place the pallet & the node in any specific directories relative to one another). Since your pallet is on github, why not [specify it as git dependency](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-git-repositories)?
2. Multiple warnings issued when running `cargo clippy` -> please assess their relevance and fix
3. There is no (easy) way for a user to verify the state of the blockchain other than trusting your UI or writing scripts to verify account / storage info. It would be great if a user could go to http://polkadot.js.org/apps and be able to browse & interact with the data. Could you provide a JSON with your custom types, that one could import into the Settings -> Developer section of polkadot.js/apps?
This is especially important since by default the mobile app would connect to your testnet (I don't know what code is running there), whereas the local webapp can be configured to connect to a local chain that I built and have full control over. It wasn't **difficult** to connect my mobile app to the local testnet (using http://192.168.XX.YY:9944) as well, but it's perhaps non-intuitive to users that mobile & web can be connected to different endpoints.
4. Since the account created on the mobile app can't be chosen/imported, I instead used the seed provided on mobile to import that account into my polkadot-js extension and use it as `browser_account`. So in your `send_registration_application` method, the same account would be used as `sender` as well as `browser_account`. This is probably quite an edge case, but I feel like you should be aware of it, as it's likely not the intended use :)
5. I'm not sure how this happened, your [original application](https://github.com/w3f/Open-Grants-Program/blob/master/applications/Gluon_decentralized_hardware_crypto_wallet_services.md#milestone-1-example--web-app-and-mobile-app-pairing) doesn't mention unit tests (we usually require these for every milestone). This is a fault on our side and since it's not part of your contract, I can't require you to write unit tests in order to accept this milestone... still, it would be nice to see these, as they speed up the evaluation process and also proves that your code is of highest quality :)
6. In your [readme for the gluon-app](https://github.com/tearust/gluon-pallet#functions-of-gluon-pallet), would you mind documenting the functions (used for this milestone) in a bit more descriptive manner? Perhaps it's just me, but in the beginning it wasn't clear to me how the actions I performed in the walkthrough corresponded to the methods called on the gluon-pallet.

So as you see, these are not issues related to the functionality of the pallet per-se, but rather good dev practices. Apart from these, your codebase looks pretty good. Let me know if you could review the above issues in M1. If you would rather do some of these (especially point 1: refactoring or 2: clippy warnings) later that will also be fine by me, as I understand that it might mess with your current development and be counter-productive at this stage -> in that case I'll make a note, but please be sure to fix them before submitting M2. However, for this milestone I would ask you to address 3 & 6 before I accept it.

kevingzhang commented:
@mmagician Thank you for your feedback. 
See my point to point reply below

> Missed Points:
> There is a lot of code duplication in gluon-app/mobile/layer1/[gluon, rpc, ...].js and gluon-app/webapp/src/tea/[gluon, rpc, ...].js.
Because we are in the middle of milestone 2. Making such refactoring is part of milestone 2 target, so we will refactor the code structure before submitting M2.

> There is a lot of code duplication in gluon-app/mobile/layer1/[gluon, rpc, ...].js and gluon-app/webapp/src/tea/[gluon, rpc, ...].js.
Done with commit https://github.com/tearust/tea-layer1/commit/ea5ee876c7c448a380f44601e0d7bfd2889f4b71


> Furthermore:
> 1. You're assuming that the gluon pallet is located in ../../gluon-pallet/gluon, which depends on the users' setup (I haven't seen instructions to place the pallet & the node in any specific directories relative to one another). Since your pallet is on github, why not specify it as git dependency?

Done with https://github.com/tearust/tea-layer1/commit/c644386f2e9172075f1e33ff0cf8b64562f90f9

> 2. Multiple warnings issued when running cargo clippy -> please assess their relevance and fix

Will fix before submit M2.

> 3. There is no (easy) way for a user to verify the state of the blockchain other than trusting your UI or writing scripts to verify account / storage info. It would be great if a user could go to http://polkadot.js.org/apps and be able to browse & interact with the data. Could you provide a JSON with your custom types, that one could import into the Settings -> Developer section of polkadot.js/apps?
> This is especially important since by default the mobile app would connect to your testnet (I don't know what code is running there), whereas the local webapp can be configured to connect to a local chain that I built and have full control over. It wasn't difficult to connect my mobile app to the local testnet (using http://192.168.XX.YY:9944) as well, but it's perhaps non-intuitive to users that mobile & web can be connected to different endpoints.

See the updated readme.md here
https://github.com/tearust/tea-layer1/tree/milestone-1#integrate-with-polkadot-web-wallet. 
The update commit https://github.com/tearust/tea-layer1/commit/0c059a946c28bc1ee626b876b50889d8dec92955

> 4.Since the account created on the mobile app can't be chosen/imported, I instead used the seed provided on mobile to import that account into my Polkadot-js extension and use it as browser_account. So in your send_registration_application method, the same account would be used as sender as well as browser_account. This is probably quite an edge case, but I feel like you should be aware of it, as it's likely not the intended use :)

We will fix this issue before submitting M2.

> 5. I'm not sure how this happened, your original application doesn't mention unit tests (we usually require these for every milestone). This is a fault on our side and since it's not part of your contract, I can't require you to write unit tests in order to accept this milestone... still, it would be nice to see these, as they speed up the evaluation process and also proves that your code is of highest quality :)

No matter you require or not, we will need unit test for our own testing purpose. 
For tea and gluon pallets, update unit-test on commits
https://github.com/tearust/tea-layer1/commit/07c3f6ed3c87560527f3393e1b46b70663afba99, https://github.com/tearust/gluon-pallet/commit/5dd303325be55a339efafe231707a4ed5a545ac9. 

> 6. In your readme for the gluon-app, would you mind documenting the functions (used for this milestone) in a bit more descriptive manner? Perhaps it's just me, but in the beginning it wasn't clear to me how the actions I performed in the walkthrough corresponded to the methods called on the gluon-pallet.

Check this out
https://github.com/tearust/gluon-app/tree/milestone1#workflow-description. 
Updated the workflow description for M1 on commit https://github.com/tearust/gluon-app/commit/33b23c23f81aaf3c950f6575d4ea1813717247dd.


Please let us know if we understood incorrectly or missed any pieces. We probably will submit next milestone very soon with all these missing pieces. 

Thanks.

mmagician commented:
@kevingzhang My apologies, I haven't seen the types.json file! I was now able to follow the steps in your tutorial and verify that the correct events are indeed dispatched.

Just to clarify: the currently present [unit tests](https://github.com/tearust/gluon-pallet/blob/master/gluon/src/tests.rs) don't actually test the functionality from M1, but rather some signing/verifying with different curves - I guess you only added the instructions as a placeholder for the future, right?

Thanks for updating the READMEs, the rest looks good and I'll be looking forward to seeing added functionality in the future milestones!

mmagician commented:
@kevingzhang The invoice you submitted is for the entire amount of the grant. Please submit a new one for the value of just the first milestone.

kevingzhang commented:
@mmagician oh yeah, I just updated the invoice with the first milestone payment amount in a new invoice google form. Please let me know if it doesn't work.

mmagician commented:
@kevingzhang Great, I've received it, thanks!

kevingzhang commented:
Hi Marcin,
May I ask when are we supposed to release our open grant to the public
given our milestone 1 is accepted (I assume it is)?

On Thu, Apr 8, 2021 at 12:59 AM Marcin ***@***.***> wrote:

> @kevingzhang <https://github.com/kevingzhang> Great, I've received it,
> thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/w3f/Grant-Milestone-Delivery/pull/125#issuecomment-815543894>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANOEEOIUP2STD4IPDIBVC3THVO6FANCNFSM4ZWRHZMA>
> .
>


-- 
╭︿︿︿╮
{/ o  o /}
 ( (oo) )     Kevin G. Zhang  :)
  ︶ ︶︶

mmagician commented:
Do you mean making an announcement?
If so, please follow the guidelines specified in [this document](https://github.com/w3f/General-Grants-Program/blob/master/grants/announcement-guidelines.md).

kevingzhang commented:
@mmagician , thank you for closing this PR. Does it mean the milestone-1 is done? Is there anything missing on our side before the milestone1 is completed? If nothing else, when do we expect to receive the payment for milestone-1, we are very close to sending PR for the milestone-2.

mmagician commented:
@kevingzhang There's nothing else to be done on your side for M1. 
I have submitted your invoice to our finance team a couple of days ago. You should receive the payment once that is processes, it might take a few days.
